lib/_stream_readable.js

```diff
@@ -839,68 +826,107 @@ Readable._fromList = fromList;
+    ret = state.buffer.shift();
+  else if (!n || n >= state.length) {
+    // read it all, truncate the list
+    if (state.decoder)
+      ret = state.buffer.join('');
+    else if (state.buffer.length === 1)
+      ret = state.buffer.head.data;
+    else
+      ret = state.buffer.concat(state.length);
+    state.buffer.clear();
+  } else {
+    // read part of list
+    ret = fromListPartial(n, state.buffer, state.decoder);
+  }
 
-  // nothing in the list, definitely empty.
-  if (list.length === 0)
-    return null;
+  return ret;
+}
 
-  if (length === 0)
-    ret = null;
-  else if (objectMode)
+function fromListPartial(n, list, hasStrings) {
+  var ret;
+  if (n < list.head.data.length) {
+    // slice is the same for buffers and strings
+    ret = list.head.data.slice(0, n);
+    list.head.data = list.head.data.slice(n);
+  } else if (n === list.head.data.length) {
+    // first chunk is a perfect match
     ret = list.shift();
-  else if (!n || n >= length) {
-    // read it all, truncate the array.
-    if (stringMode)
-      ret = list.join('');
-    else if (list.length === 1)
-      ret = list[0];
-    else
-      ret = Buffer.concat(list, length);
-    list.length = 0;
   } else {
-    // read just some of it.
-    if (n < list[0].length) {
-      // just take a part of the first list item.
-      // slice is the same for buffers and strings.
-      const buf = list[0];
-      ret = buf.slice(0, n);
-      list[0] = buf.slice(n);
-    } else if (n === list[0].length) {
-      // first list is a perfect match
-      ret = list.shift();
-    } else {
-      // complex case.
-      // we have enough to cover it, but it spans past the first buffer.
-      if (stringMode)
-        ret = '';
-      else
-        ret = Buffer.allocUnsafe(n);
-
-      var c = 0;
-      for (var i = 0, l = list.length; i < l && c < n; i++) {
-        const buf = list[0];
-        var cpy = Math.min(n - c, buf.length);
-
-        if (stringMode)
-          ret += buf.slice(0, cpy);
-        else
-          buf.copy(ret, c, 0, cpy);
+    // result spans more than one buffer
+    ret = (hasStrings
+           ? copyFromBufferString(n, list)
+           : copyFromBuffer(n, list));
+  }
+  return ret;
+}
 
-        if (cpy < buf.length)
-          list[0] = buf.slice(cpy);
+function copyFromBufferString(n, list) {
+  var p = list.head;
+  var c = 1;
+  var ret = p.data;
+  n -= ret.length;
+  while (p = p.next) {
+    const str = p.data;
+    const nb = (n > str.length ? str.length : n);
+    if (nb === str.length)
+      ret += str;
+    else
+      ret += str.slice(0, n);
+    n -= nb;
+    if (n === 0) {
+      if (nb === str.length) {
+        ++c;
+        if (p.next)
+          list.head = p.next;
         else
-          list.shift();
-
-        c += cpy;
+          list.head = list.tail = null;
+      } else {
+        list.head = p;
+        p.data = str.slice(nb);
       }
+      break;
     }
+    ++c;
   }
+  list.length -= c;
+  return ret;
+}
```

Comment:
Functions like this, while I'm sure they perform great, look incredibly hard to maintain (and review). Some comments would be useful.